p2p/src/sync/mod.rs

```diff
@@ -89,445 +97,484 @@ where
-        tx_peer_manager: mpsc::UnboundedSender<PeerManagerEvent<T>>,
+        peer_event_receiver: mpsc::UnboundedReceiver<SyncControlEvent<T>>,
+        peer_manager_sender: mpsc::UnboundedSender<PeerManagerEvent<T>>,
     ) -> Self {
         Self {
-            chain_config,
-            _p2p_config: p2p_config,
-            peer_sync_handle: handle,
-            rx_sync,
-            tx_peer_manager,
+            _chain_config: chain_config,
+            p2p_config,
+            messaging_handle,
+            peer_event_receiver,
+            peer_manager_sender,
             chainstate_handle,
             peers: Default::default(),
+            blocks_queue: Default::default(),
         }
     }
 
-    /// Get mutable reference to the handle
-    pub fn handle_mut(&mut self) -> &mut T::SyncingMessagingHandle {
-        // TODO: get rid of this function as it's used only in tests; perhaps a better way to do this is by
-        // creating p2p objects and make them communicate together instead of having access to internal
-        // private parts of the sync manager
-        &mut self.peer_sync_handle
-    }
+    /// Runs the sync manager event loop.
+    pub async fn run(&mut self) -> Result<Void> {
+        log::info!("Starting SyncManager");
 
-    /// Register peer to the `SyncManager`
-    pub async fn register_peer(&mut self, peer_id: T::PeerId) -> crate::Result<()> {
-        ensure!(
-            !self.peers.contains_key(&peer_id),
-            P2pError::PeerError(PeerError::PeerAlreadyExists),
-        );
+        let mut new_tip_receiver = self.subscribe_to_new_tip().await?;
+        loop {
+            tokio::select! {
+                event = self.messaging_handle.poll_next() => match event? {
+                    SyncingEvent::Message { peer, message } => {
+                        let res = self.handle_message(peer, message).await;
+                        self.handle_result(peer, res).await?;
+                    },
+                    SyncingEvent::Announcement{ peer, announcement } => {
+                        let res = self.handle_announcement(peer, announcement).await;
+                        self.handle_result(peer, res).await?;
+                    }
+                },
+                event = self.peer_event_receiver.recv() => match event.ok_or(P2pError::ChannelClosed)? {
+                    SyncControlEvent::Connected(peer_id) => self.register_peer(peer_id).await?,
+                    SyncControlEvent::Disconnected(peer_id) => self.unregister_peer(peer_id),
+                },
+                block_id = new_tip_receiver.recv(), if !self.chainstate_handle.call(|c| c.is_initial_block_download()).await?? => {
+                    // This error can only occur when chainstate drops an events subscriber.
+                    let block_id = block_id.ok_or(P2pError::ChannelClosed)?;
+                    self.handle_new_tip(block_id).await?;
+                },
```

Comment:
I think this line will slow down the event loop a lot, because it will block processing until the `chainstate` call returns. I think it would be much better to somehow listen for `is_initial_block_download` updates instead.

Reply:
OK I now see the context of that message on slack. 

We don't need to do that. The answer is much simpler: You can always check against the best block known from the last announcement, then check only once if that changed. You can even cache the result on every announcement and just load it here.

Reply:
I think that I have fixed this issue.